tests/system/Database/BaseQueryTest.php

```diff
@@ -420,6 +420,163 @@ public function testSetQueryBinds()
+
+        $expected = "SELECT * FROM users WHERE name = 'John' OR id = 1";
+
+        $this->assertSame($expected, $query->getQuery());
+    }
+
+    public function testSetQueryMultipleTimesReplacingBinds()
+    {
+        $query = new Query($this->db);
+
+        $query->setQuery('SELECT * FROM users WHERE name = :name:', ['name' => 'John']);
+
+        // Triggers compileBinds()
+        $query->getQuery();
+
+        $query->setQuery('SELECT * FROM users WHERE name = :name: OR id = 1', ['name' => 'Jane']);
+
+        $expected = "SELECT * FROM users WHERE name = 'Jane' OR id = 1";
+
+        $this->assertSame($expected, $query->getQuery());
+    }
+
+    public function testSetQueryMultipleTimesRemovingBinds()
+    {
+        $query = new Query($this->db);
+
+        $query->setQuery('SELECT * FROM users WHERE name = :name:', ['name' => 'John']);
+
+        // Triggers compileBinds()
+        $query->getQuery();
+
+        $query->setQuery('SELECT * FROM users WHERE name = :name: OR id = 1', []);
+
+        $expected = 'SELECT * FROM users WHERE name = :name: OR id = 1';
+
+        $this->assertSame($expected, $query->getQuery());
+    }
+
+    public function testSwapPrefixMultipleTimes()
+    {
+        $query = new Query($this->db);
+
+        $origPrefix = 'db_';
+        $newPrefix1 = 'ci_';
+        $newPrefix2 = 'zz_';
+
+        $origSQL = 'SELECT * FROM db_users WHERE db_users.id = 1';
+        $newSQL  = 'SELECT * FROM zz_users WHERE zz_users.id = 1';
+
+        $query->setQuery($origSQL);
+        $query->swapPrefix($origPrefix, $newPrefix1);
+        $query->swapPrefix($newPrefix1, $newPrefix2);
+
+        $this->assertSame($newSQL, $query->getQuery());
+    }
+
+    public function testSwapPrefixBeforeSetBinds()
+    {
+        $query = new Query($this->db);
+
+        $origSQL = 'SELECT * FROM db_users WHERE db_users.name = ?';
+        $newSQL  = "SELECT * FROM ci_users WHERE ci_users.name = 'John'";
+
+        $query->setQuery($origSQL);
+
+        $query->swapPrefix('db_', 'ci_');
+        $query->setBinds(['John']);
+
+        $this->assertSame($newSQL, $query->getQuery());
+    }
+
+    public function testSwapPrefixAfterSetBinds()
+    {
+        $query = new Query($this->db);
+
+        $origSQL = 'SELECT * FROM db_users WHERE db_users.name = ?';
+        $newSQL  = "SELECT * FROM ci_users WHERE ci_users.name = 'John'";
+
+        $query->setQuery($origSQL);
+
+        $query->setBinds(['John']);
+        $query->swapPrefix('db_', 'ci_');
+
+        $this->assertSame($newSQL, $query->getQuery());
+    }
+
+    public function testSwapPrefixIsAppliedBeforeBinds()
+    {
+        $query = new Query($this->db);
+
+        $origSQL = 'SELECT * FROM db_users WHERE db_users.name = ?';
+        $newSQL  = "SELECT * FROM ci_users WHERE ci_users.name = 'John db_foobar John'";
+
+        $query->setQuery($origSQL);
+
+        $query->setBinds(['John db_foobar John']);
+        $query->swapPrefix('db_', 'ci_');
```

Comment:
```suggestion
        $query->swapPrefix('db_', 'ci_');
        $query->setBinds(['John db_foobar John']);
```

Reply:
No, the purpose of this test is precisely to ensure that if we call, in this order, `setBinds()` then `swapPrefix()`, the binds are still applied *after* the prefix swap.

The reason is that [`swapPrefix()`](https://github.com/codeigniter4/CodeIgniter4/blob/4e3792d1382518c2f6f89d03aab857356c45a040/system/Database/Query.php#L252-L259) is quite dumb, and just operates on the whole query string. If it were applied after the binds, it would also replace in the binded values, as demonstrated in this test.

As I listed in the PR description:
> - `swapPrefix()` will always be executed on the original SQL string (with the bind placeholders), which is a bit safer than on the final SQL string (as the search/replace just operates on the whole SQL string).

Reply:
Okay, I got it.